clients/da-vinci-client/src/main/java/com/linkedin/davinci/blob/BlobSnapshotManager.java

```diff
@@ -0,0 +1,131 @@
+    File partitionSnapshotDir = new File(fullPathForPartitionDBSnapshot);
+    if (partitionSnapshotDir.exists()) {
+      partitionSnapshotDir.delete();
+      return;
+    }
+    try {
+      Checkpoint checkpoint = createCheckpoint(rocksDB);
+
+      LOGGER.info("Creating snapshots in directory: {}", fullPathForPartitionDBSnapshot);
+      checkpoint.createCheckpoint(fullPathForPartitionDBSnapshot);
+
+      LOGGER.info("Finished creating snapshots in directory: {}", fullPathForPartitionDBSnapshot);
+    } catch (RocksDBException e) {
+      throw new VeniceException(
+          "Received exception during RocksDB's snapshot creation in directory " + fullPathForPartitionDBSnapshot,
+          e);
+    }
+  }
+
+  /**
+   * Checks if the snapshot is stale, if it is and no one is using it, it updates the snapshot,
+   * otherwise it increases the count of people using the snapshot
+   */
+  public void maybeUpdateHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
+    if (rocksDB == null || topicName == null) {
+      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
+    }
+    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
+    snapShotTimestamps.putIfAbsent(topicName, new VeniceConcurrentHashMap<>());
+    concurrentSnapshotUsers.putIfAbsent(topicName, new VeniceConcurrentHashMap<>());
+
+    if (isSnapshotStale(topicName, partitionId)) {
+      if (concurrentSnapshotUsers.get(topicName).get(partitionId) == null
+          || concurrentSnapshotUsers.get(topicName).get(partitionId).get() == 0) {
+        updateHybridSnapshot(rocksDB, topicName, partitionId);
+        snapShotTimestamps.get(topicName).put(partitionId, System.currentTimeMillis());
+      }
+      if (getConcurrentUsers(rocksDB, topicName, partitionId) == 0) {
+        concurrentSnapshotUsers.get(topicName).put(partitionId, new AtomicLong(1));
+      } else {
+        concurrentSnapshotUsers.get(topicName).get(partitionId).incrementAndGet();
+      }
+      return;
+    }
+    try {
+      if (concurrentSnapshotUsers.get(topicName).get(partitionId) != null
+          && concurrentSnapshotUsers.get(topicName).get(partitionId).get() >= 0) {
+        concurrentSnapshotUsers.get(topicName).get(partitionId).incrementAndGet();
+      } else {
+        concurrentSnapshotUsers.get(topicName).put(partitionId, new AtomicLong(1));
+      }
+      LOGGER.info("Retrieved preexisting snapshot in directory: {}", fullPathForPartitionDBSnapshot);
+    } catch (Exception e) {
+      throw new VeniceException("Error getting snapshot", e);
+    }
+  }
+
+  public void decreaseConcurrentUserCount(String topicName, int partitionID) {
+    if (concurrentSnapshotUsers.containsKey(topicName)
+        && concurrentSnapshotUsers.get(topicName).containsKey(partitionID)) {
+      if (concurrentSnapshotUsers.get(topicName).get(partitionID).get() > 0) {
```

Comment:
If the logic is correct, we should not call this before we increment the count.
Maybe can you throw a VeniceException if the result is below 0?

Also, a format suggestion is, can we do this to avoid wrapping layers of if:
` someMap = concurrentSnapshotUsers.get(topicName);
 if (someMap == null) {
  throw new VeniceException("xxxx");
 }
 someAtomicLong = someMap.get(partitionId);
 if (someAtomicLong == null) {
  throw new VeniceException("xxxx");
 }
 result = someAtomicLong.decrementAndGet();
 if (result < 0) {
  throw new VeniceException("xxxx");
 }

`

Reply:
implemented that logic